cores/esp8266/Stream.h

```diff
@@ -102,12 +113,104 @@ class Stream: public Print {
         virtual String readString();
         String readStringUntil(char terminator);
 
+        virtual int read (uint8_t* buffer, size_t len);
+        int read (char* buffer, size_t len) { return read((uint8_t*)buffer, len); }
+
+        //////////////////// extension: direct access to input buffer
+        // to provide when possible a pointer to available data for read
+
+        // informs user and ::to*() on effective buffered peek API implementation
+        // by default: not available
+        virtual bool peekBufferAPI () const { return false; }
+
+        // returns number of byte accessible by peekBuffer()
+        virtual size_t peekAvailable () { return 0; }
+
+        // returns a pointer to available data buffer (size = peekAvailable())
+        // semantic forbids any kind of ::read()
+        //     - after calling peekBuffer()
+        //     - and before calling peekConsume()
+        virtual const char* peekBuffer () { return nullptr; }
+
+        // consumes bytes after peekBuffer() use
+        // (then ::read() is allowed)
+        virtual void peekConsume (size_t consume) { (void)consume; }
+
+        // by default read timeout is possible (incoming data from network,serial..)
+        // children can override to false (like String::)
+        virtual bool inputTimeoutPossible () { return true; }
+
+        // (outputTimeoutPossible() is defined in Print::)
+
+        //////////////////// extensions: Streaming streams to streams
+        // Stream::to*()
+        //
+        // Stream::to*() uses 1-copy transfers when peekBuffer API is
+        // available, or makes a regular transfer through a temporary buffer.
+        //
+        // - for efficiency, Stream classes should implement peekAPI when
+        //   possible
+        // - for an efficient timeout management, Print/Stream classes
+        //   should implement {output,input}TimeoutPossible()
+
+        using oneShotMs = esp8266::polledTimeout::oneShotFastMs;
+
+        // ::to*() methods:
+        // - always stop before timeout when "no-more-input-possible-data"
+        //   or "no-more-output-possible-data" condition is met
+        // - always return number of transfered bytes
+        // When result is 0 or less than requested maxLen, this->getLastTo()
+        // contains an error reason.
+
+        // transfers already buffered / immediately available data (no timeout)
+        // returns number of transfered bytes
+        size_t sendNow (Print* to) { return sendGeneric(to, -1, -1, oneShotMs::alwaysExpired); }
+        size_t sendNow (Print& to) { return sendNow(&to); }
+
+        // transfers data until timeout
```

Comment:
> What exactly times out?

Here is the current documentation added in the PR for `::sendAll()`:

    - ``Stream::sendAll(dest [, timeout])``

      This method waits up to the given or default timeout to transfer all
      available data.  It is useful when source is able to tell that no more
      data will be available for this call, or when destination is able to
      tell that it will no more be able to receive.

      For example, a source String will not grow during the transfer, or a
      particular network connection supposed to send a fixed amount of data
      before closing.  ``::sendAll()`` will receive all bytes.  Timeout is
      useful when destination needs processing time (e.g.  network or serial
      input buffer full = please wait a bit).

I wonder whether that explanation is clear enough.